drivers/audio/csaudiork3x/Source/Main/NewDelete.cpp

```diff
@@ -0,0 +1,141 @@
+/*****************************************************************************
+* NewDelete.cpp -  CPP placement new and delete operators implementation
+*****************************************************************************
+* Copyright (c) Microsoft Corporation All Rights Reserved
+*
+* Module Name:
+*
+* NewDelete.cpp
+*
+* Abstract:
+*
+*   Definition of placement new and delete operators.
+*
+*/
+
+#ifdef _NEW_DELETE_OPERATORS_
+#ifdef __cplusplus
+extern "C" {
+#include <wdm.h>
+}
+#else
+#include <wdm.h>
+#endif
+
+#include "newDelete.h"
+#include "definitions.h"
+
+#pragma code_seg()
+/*****************************************************************************
+* Functions
+*/
+
+/*****************************************************************************
+* ::new()
+*****************************************************************************
+* New function for creating objects with a specified allocation tag.
+*/
+PVOID operator new
+(
+    size_t      iSize,
+    POOL_TYPE   poolType,
+    ULONG       tag
+)
+{
+    PVOID result = ExAllocatePoolZero(poolType, iSize, tag);
+
+    return result;
+}
+
+
+/*****************************************************************************
+* ::new()
+*****************************************************************************
+* New function for creating objects with a specified allocation tag.
+*/
+PVOID operator new
+(
+    size_t      iSize,
+    POOL_TYPE   poolType
+)
+{
+    PVOID result = ExAllocatePoolZero(poolType, iSize, CSAUDIORK3X_POOLTAG);
+
+    return result;
+}
+
+
+/*****************************************************************************
+* ::delete()
+*****************************************************************************
+* Delete with tag function.
+*/
+void __cdecl operator delete
+(
+    PVOID pVoid,
+    ULONG tag
+)
+{
+    if (pVoid)
+    {
+        ExFreePoolWithTag(pVoid, tag);
+    }
+}
+
+
+/*****************************************************************************
+* ::delete()
+*****************************************************************************
+* Sized Delete function.
+*/
+void __cdecl operator delete
+(
+    _Pre_maybenull_ __drv_freesMem(Mem) PVOID pVoid,
+    _In_ size_t cbSize
+)
+{
+    UNREFERENCED_PARAMETER(cbSize);
+
+    if (pVoid)
+    {
+        ExFreePoolWithTag(pVoid, CSAUDIORK3X_POOLTAG);
```

Comment:
Probably should just be ExFreePool since the memory could have been allocated with a different tag. (ExFreePoolWithTag is usually just a macro that calls ExFreePool.)

Reply:
Microsoft is using ExFreePoolWithTag in their sample (which this is based off) so it seems they're confident it's from the same tag https://github.com/microsoft/Windows-driver-samples/blob/main/audio/simpleaudiosample/Source/Main/NewDelete.cpp#L121

Reply:
Current implementation of ExFreePoolWithTag doesn't actually check the tag parameter at all. It's possible that this changes if verifier is enabled.

Using ExFreePoolWithTag just means "I claim that I know what tag was used here". If you don't actually know that (i.e. if your `operator delete` can be paired with an `operator` new that takes a tag parameter) then it seems weird to go to extra effort to claim that.